### ISSUE TITLE
Request: ordinal logistic loss

### ISSUE THREAD

**Issue by david-cortes** [catboost/catboost#1994]
Would be good to have the equivalent of a typical ordinal logistic regression model as a possible loss for multi-class classification: https://en.wikipedia.org/wiki/Ordinal_regression

In short, this is a model which estimates the log-odds based on setting thresholds for each category on a single raw prediction value, and I guess it's one of the most commonly used models for tabular data.

Here's an article explaining it in a bit more detail: http://www.erikdrysdale.com/ordinal_regression/

It differs a bit from the current logic for multi-class classification as there's a couple more parameters being optimized (namely, the thresholds) on which the regularization should not apply, and there's a single raw value per instance instead of one per class like in multinomial regression, but I guess it should be possible to accomodate it within the boosting logic.

**Comment by jasonsross** [comment]
I think this would be a powerful thing to implement for catboost. Even sklearn does not have an in-house ordinal classifier or estimator wrapper. There's some toy built classifiers and wrappers out there in personal githubs but it's hard to parse through the quality of their work as there's several pitfalls to avoid with ordinal classification techniques.

**Comment by adamingas** [comment]
I've written a loss (with gradient and hessian) and a lightGBM model wrapper that learns the a latent variable which is then used to calculate the log-odds and probabilities for each level. I will be writing CatBoost and XGBoost implementations of it soon.

Model building happens in two steps
1) Build a small number of trees using thresholds equidistant from each other
2) Produce predictions and hold them constant while optimising the loss with respect to the thresholds

This is repeated once in my current implementation, and I call it hot start.

You can find more information in the [repo](https://github.com/adamingas/ordinalgbt/) and you can play around with it by installing it with `pip install ordinalgbt`. Contributions are welcomed. 

Let me know if this is the wrong place to write about this and my comment needs deleting.

**Comment by david-cortes** [comment]
Thanks for sharing!

For initialization, the optimal thresholds (given that predictions are all zeros) can be gotten in closed-form from the frequencies of the classes though, which might provide some improvement compared to a constant initialization.
